modules/491-containerized-data-importer/hooks/create_registry_secret_for_cdi.go

```diff
@@ -0,0 +1,187 @@
+			NameSelector: &types.NameSelector{
+				MatchNames: []string{"deckhouse-registry"},
+			},
+			NamespaceSelector: &types.NamespaceSelector{
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{"d8-system"},
+				},
+			},
+			FilterFunc: applyRegistrySecretFilter,
+		},
+	},
+}, handleCDIPod)
+
+func applyCDIPodFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	return obj.GetNamespace(), nil
+}
+
+func applyRegistrySecretFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var s corev1.Secret
+	err := sdk.FromUnstructured(obj, &s)
+	if err != nil {
+		return "", err
+	}
+
+	ns := s.GetNamespace()
+
+	conf, ok := s.Data[corev1.DockerConfigJsonKey]
+	if !ok {
+		return "", fmt.Errorf("registry auth conf is not in registry secret %s/%s", ns, obj.GetName())
+	}
+
+	return registrySecret{
+		Namespace: ns,
+		Config:    string(conf),
+	}, nil
+}
+
+func prepareVirtRegistrySecret(namespace, dockerCfg string) *corev1.Secret {
+	return &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      virtRegistrySecretName,
+			Namespace: namespace,
+			Labels: map[string]string{
+				"heritage": "deckhouse",
+				"kubevirt.deckhouse.io/cdi-registry-secret": "true",
+				"deckhouse.io/registry-secret":              "true",
+			},
+		},
+
+		Data: map[string][]byte{
+			corev1.DockerConfigJsonKey: []byte(dockerCfg),
+		},
+
+		Type: corev1.SecretTypeDockerConfigJson,
+	}
+}
+
+// handleCDIPod
+//
+// synopsis:
+//
+//	Every namespace running cdi workload pod, must contain registry
+//	secret to be able pulling cdi images.
+//	The cdi images re-pushed in deckhouse registry to make
+//	them possible to run in closed environments.
+//	After CDI pod is finished the registry secret must be deleted.
+//
+//	We have patched cdi-controller for specifying image pull-secrets
+//	for envery VMI pod. This is temproray solution, until kubevirt
+//	will have native opportunity for specifying registrySecrets.
+func handleCDIPod(input *go_hook.HookInput) error {
+	d8RegistrySnap := input.Snapshots[d8RegistrySnapshot]
+	if len(d8RegistrySnap) == 0 {
+		input.LogEntry.Warnln("Registry secret not found. Skip")
+		return nil
+	}
+
+	registryCfg := d8RegistrySnap[0].(registrySecret).Config
+
+	cdiPods := set.NewFromSnapshot(input.Snapshots[cdiPodsSnapshot])
+
+	// namespace -> .dockerconfigjson content
+	secretsByNs := map[string]string{}
+
+	for _, sRaw := range input.Snapshots[secretsSnapshot] {
+		regSecret := sRaw.(registrySecret)
+		secretsByNs[regSecret.Namespace] = regSecret.Config
+	}
+
+	// create secrets
+	for ns := range cdiPods {
+		secretContent, secretExists := secretsByNs[ns]
+		// secret already exists in namespace. do not create or patch
+		if !secretExists || secretContent != registryCfg {
+			secret := prepareVirtRegistrySecret(ns, registryCfg)
+			input.PatchCollector.Create(secret, object_patch.UpdateIfExists())
```

Comment:
There is helm templates, why do you create the secret from hook?

Reply:
This is same as for cert-manager, we don't know how soon and how many resources user will create

Reply:
Please, mark in "phase 2" to reorganize this. I believe, the cert-manager decision was also wrong.
Or if it isn't a deal, do it in this pr.

Reply:
Internally decided to do this by hooks. I added task to backlog to implement this on KubeVirt and CDI sides